Fix CLASSIC_JERK to consistently respect jerk limits

mh-dm opened the pull request:
### Description
This fixes 3 bugs/issues in jerk limiting:
1) Computed v_entry speed was wrong (missing a 'current_speed_factor') when the current block nominal speed was > previous nominal speed. Thus the resulting junction speed was wrong.
2) When switching directions (say nearly 180 deg) the computed jerk would be completely overridden by applying separate 'safe' exit / entry speeds. This results in basically doubling the jerk limit for sharp back and forth movements. Presumably this was done to address slowdowns, but after fixing 1) this special handling is no longer needed.
3) Jerk is a change in speed (i.e. ABS(v_exit - v_entry)) and not _MAX(v_exit, -v_entry), even in axis reversal.

Other notes:
- Less special casing, shorter, cleaner code.
- Slightly faster jerk computation as we stop having to always compute the 'safe' speed (jerk to halt speed).
- Refactored the code that computes the 'safe' speed.

### Benefits

The overall result is smoother motion when printing models that involve lots of small back and forth moves, like on thin strips in top layers. Alternatively, jerk limits can be slightly increased for faster (or sometimes higher quality) prints while keeping overall vibrations at same level.

Reduced chance of surprise layer shifts. Say you tune jerk and acceleration using some non-extreme model (say a calibration cube). Later, when printing something with a lot of sharp back and forth movements you get layer shifts.

### Configurations

Affects using CLASSIC_JERK. As part of a journey to getting an Ender-3 to print really fast I've tested (on top of) this change quite a bit.

### Related Issues

A bit hard find related issues as most people would tune down jerk and acceleration if the movement sounded too jerky and acceleration defaults are often conservative. I did find one case of surprise Layer shifting #23800 that mentioned "Y axis only has very small movements".

mh-dm committed:
Fix CLASSIC_JERK to consistently respect jerk limits.

This fixes 3 bugs/issues in jerk limiting:
1) Computed v_entry speed was wrong (missing a 'current_speed_factor') when the current block nominal speed was > previous nominal speed. Thus the resulting junction speed was wrong.
2) When switching directions (say nearly 180 deg) the computed jerk would be completely overridden by applying separate 'safe' exit / entry speeds. This results in basically doubling the jerk limit for sharp back and forth movements. Presumably this was done to address slowdowns, but after fixing 1) this special handling is no longer needed.
3) Jerk is a change in speed (i.e. ABS(v_exit - v_entry)) and not _MAX(v_exit, -v_entry), even in axis reversal.

Other notes:
- Less special casing, shorter, cleaner code.
- Slightly faster jerk computation as we stop having to always compute the 'safe' speed (jerk to halt speed).
- Refactored the code that computes the 'safe' speed.

The overall result is smoother motion when printing models that involve lots of small back and forth moves, like on thin strips in top layers. Alternatively, jerk limits can be slightly increased for faster prints while keeping overall vibrations at same level.

mh-dm commented:
Here's a before/after comparison showing improvement in vibrations. I set X and Y jerk to 10mm/s (the default in marlin). X and Y acceleration are 10k and 5k respectively, with print/travel acceleration really high (20k) to make sure said limits apply. Using 100mm/s and 300mm/s print and travel speeds. (On a heavily customized Ender-3 with hardware allows said high settings.)

https://github.com/MarlinFirmware/Marlin/assets/299015/cc0e6596-2561-43ef-9337-7d9311ed6763
https://github.com/MarlinFirmware/Marlin/assets/299015/82ec2ff3-8d8f-4d04-93cd-807b52eddfa7

Test part was specifically designed/sliced to cause many small back and forth movements and to calibrate acceleration/jerk settings.

I was going to say that the print quality is exactly the same, down to the blobs on the outer walls that are there due to the random seam setting in the slicer. In the photo, before is bottom, after is top. But no, the print quality is slightly better as there are slightly fewer blobs. Note that model is sliced at high enough resolution to be processing intensive - any code speed up means the printer doesn't have to stop to process the next segment.
To make sure the jerk code is properly benchmarked, the part includes multiple rounded outer walls of varying radii which are very sensitive to the jerk setting. If not high enough (or if the junction speed is wrongly computed) the printer would have to slow down and speed up significantly for every single small segment of a curve.

![result](https://github.com/MarlinFirmware/Marlin/assets/299015/80145540-99ae-4337-a7c9-3a9b87f205cc)

thinkyhead commented:
I had to make a new PR because (for some Githubby reason) I can't push commits to tweak the formatting of this PR. So it's now moved over to #26521. The results seem good and the code is sensible, so I'll merge that pretty soon.

mh-dm commented:
Had "Allow edits and access to secrets by maintainers" off which I now changed.
I'm happy to make the formatting changes you want.
Also, per discussion in #26521, switched to = 1.0f as well and removed TERN(HAS_LINEAR_E_JERK, LOOP_NUM_AXES, LOOP_LOGICAL_AXES)(axis) as HAS_LINEAR_E_JERK is undefined when HAS_CLASSIC_JERK.

thinkyhead commented:
I'd reopen this for convenience, but GitHub won't allow it. I'll pick up your new commits from your fork and add them to the new PR shortly.

mh-dm commented:
Do these instructions work to reopen pull request? https://gist.github.com/Farhan-Haseeb/fd4cbb476de4e2931c322d7bacb75e5f
